runtime/memprof.c

```diff
@@ -156,1176 +1258,983 @@ static memprof_domain_t domain_create(caml_domain_state *caml_state)
+/* Set the suspended flag on `domain` to `s`. Has the side-effect of
+ * setting the trigger. */
 
 static void update_suspended(memprof_domain_t domain, bool s)
 {
   if (domain->current) {
     domain->current->suspended = s;
   }
-  caml_memprof_renew_minor_sample(domain->caml_state);
+  /* If we are unsuspending, set the action-pending flag if
+   * we have callbacks to run. */
   if (!s) set_action_pending_as_needed(domain);
+
+  caml_memprof_set_trigger(domain->caml_state);
+  caml_reset_young_limit(domain->caml_state);
 }
 
-/* Set the suspended flag on the current domain to `s`. */
+/* Set the suspended flag on the current domain to `s`.
+ * Has the side-effect of setting the trigger. */
 
 void caml_memprof_update_suspended(bool s) {
   update_suspended(Caml_state->memprof, s);
 }
 
-/**** Sampling procedures ****/
-
-Caml_inline bool running(memprof_domain_t domain)
-{
-  memprof_thread_t thread = domain->current;
-
-  if (thread && !thread->suspended) {
-    value config = domain->config;
-    return Running(config);
-  }
-  return false;
-}
-
-/* Renew the next sample in a domain's minor heap. Could race with
- * sampling and profile-stopping code, so do not call from another
- * domain unless the world is stopped. Must be called after each minor
- * sample and after each minor collection. In practice, this is called
- * at each minor sample, at each minor collection, and when sampling
- * is suspended and unsuspended. Extra calls do not change the
- * statistical properties of the sampling because of the
- * memorylessness of the geometric distribution. */
-
-void caml_memprof_renew_minor_sample(caml_domain_state *state)
-{
-  memprof_domain_t domain = state->memprof;
-  value *trigger = state->young_start;
-  if (running(domain)) {
-    /* set trigger based on geometric distribution */
-  }
-  CAMLassert((trigger >= state->young_start) &&
-             (trigger <= state->young_ptr));
-  state->memprof_young_trigger = trigger;
-  caml_reset_young_limit(state);
-}
+/**** Iterating over entries ****/
 
-/**** Interface with systhread. ****/
+/* Type of a function to apply to a single entry. Returns true if,
+ * following the call, the entry may have a newly-applicable
+ * callback. */
 
-CAMLexport memprof_thread_t caml_memprof_new_thread(caml_domain_state *state)
-{
-  return thread_create(state->memprof);
-}
+typedef bool (*entry_action)(entry_t, void *);
 
-CAMLexport memprof_thread_t caml_memprof_main_thread(caml_domain_state *state)
-{
-  memprof_domain_t domain = state->memprof;
-  memprof_thread_t thread = domain->threads;
+/* Type of a function to apply to an entries array after iterating
+ * over the entries. The second argument is 'young', indicating
+ * whether the iteration was just over possibly-young entries. */
 
-  /* There should currently be just one thread in this domain */
-  CAMLassert(thread);
-  CAMLassert(thread->next == NULL);
-  return thread;
-}
+typedef void (*entries_action)(entries_t, bool, void *);
```

Comment:
```suggestion
typedef void (*entries_action)(entries_t, bool young, void *);
```

Also, this parameter is never used in any `entries_action` instance.

Reply:
I've made the change but retained the parameter (for now) for future use.